File: גירסאות/24-1/run_fix.py
// סקריפט פייתון להפעלת התיקון במסד נתונים
// מתחבר למסד נתונים ומפעיל את URGENT_FIX.sql

import psycopg2
import sys
from datetime import datetime

def connect_to_db():
    """התחברות למסד הנתונים"""
    try:
        # נסה עם הגדרות ברירת מחדל נפוצות
        conn = psycopg2.connect(
            host="localhost",
            database="postgres",  # שם ברירת מחדל
            user="postgres",     # שם משתמש ברירת מחדל
            password="password"  # סיסמה ברירת מחדל
        )
        print("✅ התחברות למסד נתונים הצליחה")
        return conn
    except Exception as e:
        print(f"❌ שגיאת התחברות: {e}")
        
        # נסה עם פרטים אחרים
        try:
            conn = psycopg2.connect(
                host="localhost",
                database="scheduling_db",
                user="scheduling_user", 
                password="scheduling_pass"
            )
            print("✅ התחברות למסד נתונים הצליחה (הגדרות אלטרנטיביות)")
            return conn
        except Exception as e2:
            print(f"❌ שגיאת התחברות שנייה: {e2}")
            print("🔧 עדכן את פרטי ההתחברות בסקריפט!")
            return None

def run_fix(conn):
    """הפעלת התיקון"""
    try:
        with conn.cursor() as cur:
            print("🔧 מפעיל תיקון...")
            
            # יצירת אינדקס ייחודי
            cur.execute("""
                DROP INDEX IF EXISTS assignments_no_double_booking;
            """)
            
            cur.execute("""
                CREATE UNIQUE INDEX assignments_no_double_booking 
                ON assignments (room_id, DATE(date), start_time, end_time) 
                WHERE status = 'active';
            """)
            
            conn.commit()
            print("✅ אינדקס ייחודי נוצר בהצלחה!")
            
            # וידוא
            cur.execute("""
                SELECT indexname, indexdef 
                FROM pg_indexes 
                WHERE tablename = 'assignments' 
                    AND indexname = 'assignments_no_double_booking'
            """)
            
            result = cur.fetchone()
            if result:
                print(f"✅ וידוא: אינדקס {result[0]} קיים")
                print(f"📋 הגדרה: {result[1]}")
            else:
                print("❌ אינדקס לא נמצא!")
            
            return True
            
    except Exception as e:
        print(f"❌ שגיאה בהפעלת התיקון: {e}")
        conn.rollback()
        return False

def test_duplicate_prevention(conn):
    """בדיקת מניעת כפילויות"""
    try:
        with conn.cursor() as cur:
            print("\n🧪 בודק מניעת כפילויות...")
            
            # בדיקת כפילויות קיימות
            cur.execute("""
                SELECT 
                    room_id,
                    DATE(date) as assignment_date,
                    start_time,
                    end_time,
                    COUNT(*) as duplicate_count
                FROM assignments 
                WHERE status = 'active'
                GROUP BY room_id, DATE(date), start_time, end_time
                HAVING COUNT(*) > 1
                ORDER BY duplicate_count DESC
                LIMIT 5
            """)
            
            duplicates = cur.fetchall()
            
            if duplicates:
                print(f"📊 נמצאו {len(duplicates)} קבוצות כפילויות:")
                for dup in duplicates:
                    print(f"   - חדר {dup[0]} ב-{dup[1]} {dup[2]}-{dup[3]}: {dup[4]} כפילויות")
            else:
                print("✅ אין כפילויות קיימות!")
            
            # בדיקת סך הכל
            cur.execute("SELECT COUNT(*) FROM assignments WHERE status = 'active'")
            total = cur.fetchone()[0]
            
            cur.execute("""
                SELECT COUNT(DISTINCT room_id || DATE(date) || start_time || end_time) 
                FROM assignments 
                WHERE status = 'active'
            """)
            unique = cur.fetchone()[0]
            
            print(f"📊 סך הכל: {total} שיבוצים, {unique} סלוטים ייחודיים")
            
            if total == unique:
                print("✅ כל השיבוצים ייחודיים!")
            else:
                print(f"⚠️  יש {total - unique} כפילויות!")
            
    except Exception as e:
        print(f"❌ שגיאה בבדיקה: {e}")

def main():
    """פונקציה ראשית"""
    print("🚀 הפעלת תיקון מניעת כפילויות")
    print("=" * 50)
    
    # התחברות
    conn = connect_to_db()
    if not conn:
        print("❌ לא ניתן להתחבר למסד נתונים")
        sys.exit(1)
    
    try:
        # הפעלת התיקון
        if run_fix(conn):
            # בדיקת התוצאות
            test_duplicate_prevention(conn)
            
            print("\n🎉 תיקון הסתיים בהצלחה!")
            print("🛡️  המערכת עכשיו מוגנת מכפילויות!")
        else:
            print("❌ תיקון נכשל")
            
    except Exception as e:
        print(f"❌ שגיאה כללית: {e}")
    finally:
        conn.close()

if __name__ == "__main__":
    main()

File: גירסאות/שיבוץ ידני/run_fix.js
// סקריפט Node.js להפעלת התיקון
// מתחבר למסד נתונים ומפעיל את התיקון

const { Pool } = require('pg');

async function runFix() {
  console.log('🚀 הפעלת תיקון מניעת כפילויות');
  console.log('='.repeat(50));

  // הגדרות חיבור (נסה מספר אפשרויות)
  const configs = [
    {
      host: 'localhost',
      database: 'postgres',
      user: 'postgres', 
      password: 'password'
    },
    {
      host: 'localhost',
      database: 'scheduling_db',
      user: 'scheduling_user',
      password: 'scheduling_pass'
    },
    {
      host: 'localhost',
      database: 'your_database',
      user: 'your_user',
      password: 'your_password'
    }
  ];

  let pool = null;
  
  // נסה להתחבר עם כל ההגדרות
  for (const config of configs) {
    try {
      pool = new Pool(config);
      await pool.query('SELECT 1');
      console.log('✅ התחברות למסד נתונים הצליחה');
      break;
    } catch (error) {
      console.log(`❌ ניסיון התחברות נכשל: ${error.message}`);
      if (pool) await pool.end();
      pool = null;
    }
  }

  if (!pool) {
    console.log('❌ לא ניתן להתחבר למסד נתונים');
    console.log('🔧 עדכן את פרטי ההתחברות בסקריפט!');
    process.exit(1);
  }

  try {
    console.log('🔧 מפעיל תיקון...');
    
    // יצירת אינדקס ייחודי
    await pool.query('DROP INDEX IF EXISTS assignments_no_double_booking');
    
    await pool.query(`
      CREATE UNIQUE INDEX assignments_no_double_booking 
      ON assignments (room_id, DATE(date), start_time, end_time) 
      WHERE status = 'active'
    `);
    
    console.log('✅ אינדקס ייחודי נוצר בהצלחה!');
    
    // וידוא
    const result = await pool.query(`
      SELECT indexname, indexdef 
      FROM pg_indexes 
      WHERE tablename = 'assignments' 
        AND indexname = 'assignments_no_double_booking'
    `);
    
    if (result.rows.length > 0) {
      console.log(`✅ וידוא: אינדקס ${result.rows[0].indexname} קיים`);
      console.log(`📋 הגדרה: ${result.rows[0].indexdef}`);
    } else {
      console.log('❌ אינדקס לא נמצא!');
    }

    // בדיקת כפילויות
    console.log('\n🧪 בודק מניעת כפילויות...');
    
    const duplicates = await pool.query(`
      SELECT 
        room_id,
        DATE(date) as assignment_date,
        start_time,
        end_time,
        COUNT(*) as duplicate_count
      FROM assignments 
      WHERE status = 'active'
      GROUP BY room_id, DATE(date), start_time, end_time
      HAVING COUNT(*) > 1
      ORDER BY duplicate_count DESC
      LIMIT 5
    `);
    
    if (duplicates.rows.length > 0) {
      console.log(`📊 נמצאו ${duplicates.rows.length} קבוצות כפילויות:`);
      duplicates.rows.forEach(dup => {
        console.log(`   - חדר ${dup.room_id} ב-${dup.assignment_date} ${dup.start_time}-${dup.end_time}: ${dup.duplicate_count} כפילויות`);
      });
    } else {
      console.log('✅ אין כפילויות קיימות!');
    }
    
    // בדיקת סטטיסטיקות
    const totalResult = await pool.query("SELECT COUNT(*) FROM assignments WHERE status = 'active'");
    const uniqueResult = await pool.query(`
      SELECT COUNT(DISTINCT room_id || DATE(date) || start_time || end_time) 
      FROM assignments 
      WHERE status = 'active'
    `);
    
    const total = parseInt(totalResult.rows[0].count);
    const unique = parseInt(uniqueResult.rows[0].count);
    
    console.log(`📊 סך הכל: ${total} שיבוצים, ${unique} סלוטים ייחודיים`);
    
    if (total === unique) {
      console.log('✅ כל השיבוצים ייחודיים!');
    } else {
      console.log(`⚠️  יש ${total - unique} כפילויות!`);
    }

    console.log('\n🎉 תיקון הסתיים בהצלחה!');
    console.log('🛡️  המערכת עכשיו מוגנת מכפילויות!');
    
  } catch (error) {
    console.error('❌ שגיאה:', error.message);
  } finally {
    await pool.end();
  }
}

runFix().catch(console.error);
